sdk/preview_termination_queries.go

```diff
@@ -0,0 +1,794 @@
+package sdk
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math"
+	"strings"
+	"time"
+
+	corebig "math/big"
+
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/go-state-types/builtin/v9/miner"
+	"golang.org/x/exp/constraints"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-bitfield"
+	"github.com/filecoin-project/go-state-types/abi"
+	lotusapi "github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+	miner8 "github.com/filecoin-project/specs-actors/v8/actors/builtin/miner"
+	poolstypes "github.com/glifio/go-pools/types"
+	"github.com/glifio/go-pools/util"
+	"github.com/ipfs/go-cid"
+	cbor "github.com/ipfs/go-ipld-cbor"
+)
+
+// PreviewTerminateSector gets the burnt funds for when a single sector is terminated
+func (q *fevmQueries) PreviewTerminateSector(
+	ctx context.Context,
+	minerAddr address.Address,
+	tipset string,
+	vmHeight uint64,
+	sectorNumber uint64,
+	gasLimit uint64,
+	offchain bool,
+) (actor *types.ActorV5, totalBurn *corebig.Int, epoch abi.ChainEpoch, err error) {
+	lClient, closer, err := q.extern.ConnectLotusClient()
+	if err != nil {
+		return nil, nil, 0, err
+	}
+	defer closer()
+	api := *lClient
+
+	h, ts, err := parseTipSetAndHeight(ctx, api, tipset, vmHeight)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	// Lookup actor balance
+	actor, err = api.StateGetActor(ctx, minerAddr, ts.Key())
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	// Lookup current owner / worker
+	minerInfo, err := api.StateMinerInfo(ctx, minerAddr, ts.Key())
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	var params miner.TerminateSectorsParams
+
+	if gasLimit == 0 {
+		gasLimit = 6000000000
+	}
+
+	sectorPartition, err := api.StateSectorPartition(ctx, minerAddr, abi.SectorNumber(sectorNumber), ts.Key())
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	termination := miner.TerminationDeclaration{
+		Deadline:  sectorPartition.Deadline,
+		Partition: sectorPartition.Partition,
+		Sectors:   bitfield.NewFromSet([]uint64{sectorNumber}),
+	}
+	params = miner.TerminateSectorsParams{
+		Terminations: []miner.TerminationDeclaration{termination},
+	}
+
+	totalBurn, err = terminateSectors(ctx, *lClient, h, ts, minerAddr, minerInfo,
+		params, int64(gasLimit), offchain)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	return actor, totalBurn, h, nil
+}
+
+type terminationTask struct {
+	termination         miner.TerminationDeclaration
+	deadlineHeight      abi.ChainEpoch
+	deadlineTs          *types.TipSet
+	sectorsCount        int64
+	sampledSectorsCount int64
+}
+
+// PreviewTerminateSectors gets the burnt funds for when all the sectors are terminated
+func (q *fevmQueries) PreviewTerminateSectors(
+	ctx context.Context,
+	minerAddr address.Address,
+	tipset string,
```

Comment:
Wishlist - it would be nice if this was a proper tipset type so we can get assistance from the compiler when we mess up types

Reply:
It's re-using the tipset parsing code from the Lotus CLI so you can leave it blank, or specify '@height', or a tipset string. This could be moved into the CLI that's using this, but there's some fussy interactions.

Reply:
We'll leave this as-is for now, but implement an opinionated method for use by the ADO later that takes a tipset struct.